From 5022c4e5a2bd1483c70f2e8177eb51edd36139eb Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Wed, 31 Oct 2018 10:52:59 +0100
Subject: [PATCH 2/2] Hack: make relocation work inside libexec/git-core/ and
 bin/

In Git for Windows, there is a copy of libcurl-4.dll in
/mingw64/libexec/git-core/ (which is where git-remote-https.exe lives,
and the libcurl-4.dll needs to be there because of DLL search path
issues, to prevent any C:\Windows\system32\libcurl-4.dll from being
picked up by mistake).

The current code trying to make libcurl-4.dll relocatable, however, is
unprepared for any path but /mingw64/bin/ (and it does not even verify
that it *is* in that location).

The design of the relocation code seems to be less than conducive to a
proper fix for this issue, so let's just be content with a quick 'n
dirty workaround.

Note: as MinGit does away with libexec/git-core/, we also explicitly
handle the `bin/` directory. Technically, this should not be necessary,
as the CURL_BINDIR is supposedly being handled correctly, but it seems
not to work as advertised in practice. So let's just paper over this and
extend the Git for Windows-specific hack with a hard-code `bin/` case.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 lib/vtls/vtls.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/lib/vtls/vtls.c b/lib/vtls/vtls.c
index bb45551b0..a864fb758 100644
--- a/lib/vtls/vtls.c
+++ b/lib/vtls/vtls.c
@@ -343,10 +343,29 @@ CURLcode Curl_ssl_easy_config_complete(struct Curl_easy *data)
     get_dll_path(relocated_bundle, path_max);
     strip_n_suffix_folders(relocated_bundle, 1);
     strncat(relocated_bundle, "/", path_max - 1);
-    char *relative = get_relative_path(CURL_BINDIR, CURL_CA_BUNDLE);
+    char *relative;
+    size_t len;
+    if((len = strlen(relocated_bundle)) > 18 &&
+       !strcmp(relocated_bundle + len - 18, "/libexec/git-core/")) {
+      /* Hack for Git for Windows */
+      relative = strdup("../../etc/ssl/certs/ca-bundle.crt");
+    } else if(len > 5 && !strcmp(relocated_bundle + len - 5, "/bin/")) {
+      /* Hack for Git for Windows */
+      relative = strdup("../etc/ssl/certs/ca-bundle.crt");
+    } else
+      relative = get_relative_path(CURL_BINDIR, CURL_CA_BUNDLE);
     strncat(relocated_bundle, relative, path_max - 1);
     free((void*)relative);
     simplify_path(relocated_bundle);
+    if (access(relocated_bundle, R_OK) != 0 &&
+        (len = strlen(relocated_bundle)) >= 28 &&
+        !strcmp(relocated_bundle + len - 28, "/etc/ssl/certs/ca-bundle.crt")) {
+      /* Try without /etc/ */
+      memmove(relocated_bundle + len - 28, relocated_bundle + len - 24, 25);
+      if (access(relocated_bundle, R_OK) != 0)
+        /* fall back to using /etc/ */
+        strcat(relocated_bundle + len - 28, "/etc/ssl/certs/ca-bundle.crt");
+    }
 #endif /* defined(__MINGW32__) */
     if(!sslc->custom_cafile && !set->str[STRING_SSL_CAFILE]) {
 #ifdef __MINGW32__
